File: Ehsan/Lab6/non_restoring_division.py
def shiftLeft(A, Q):
    temp = (A << 32) | (Q & 0xffffffff)
    temp = temp << 1
    A = (temp >> 32) & 0xffffffff
    Q = temp & 0xffffffff
    return (A, Q)

def division(Q,M):
    if (M == 0):
        return None
    N = 32
    A = 0
    for i in range(0, N, 1):
        (A, Q) = shiftLeft(A, Q)

        if A & 0x80000000 == 0:  
            A = (A - M) & 0xffffffff
        else:
            A = (A + M) & 0xffffffff

        if A & 0x80000000 == 0:  
            Q = Q | 0x00000001 
        else:
            Q = Q & 0xfffffffe

    if A & 0x80000000 != 0:  
        A = (A + M) 
        A = A & 0xffffffff  

    return (Q,A)


Q = int(input("Enter dividend : "))
M = int(input("Enter divisor : "))

result = division(Q, M)

if result is None:
    print("Division by zero is not allowed.")
else:
    quotient, remainder = result
    print("Quotient:", quotient)
    print("Remainder:", remainder)

File: Nasir/src/exp5/tasks/nonRestoringDivision.c
// including header files
#include<stdio.h>
#include<stdlib.h>
#include<stdint.h>
#include<time.h>

// function prototypes
uint64_t shiftLeft(uint32_t remainder, uint32_t dividend);

void nonRestoringDivision(uint32_t dividend, uint32_t divisor, uint32_t *remainder, uint32_t *quotient);

uint64_t shiftLeft(uint32_t remainder, uint32_t dividend){
    uint64_t result;
    result = (uint64_t)remainder << 32;
    result = result | dividend;
    result <<= 1;
    return result;
}

void nonRestoringDivision(uint32_t dividend, uint32_t divisor, uint32_t *remainder, uint32_t *quotient){
    if (divisor == 0){
        return;
    }
    for (int shift = 0; shift < 32; shift++){
        // extracting remainder sign bit
        int remainderSignBit = *remainder & (1 << 31);
        // shift remainder and dividend by 1
        uint64_t remainderDividendComb = shiftLeft(*remainder, dividend);
        // update remainder
        *remainder = (uint32_t)(remainderDividendComb >> 32);
        // update dividend
        dividend = (uint32_t)(remainderDividendComb & 0xFFFFFFFF);
        // checking sign of remainder
        // if 1 a = a+m = r = r+divisor
        if (remainderSignBit) {
            *remainder += divisor;
        }
        else{
            *remainder -= divisor;
        }

        // checking signbit if 1 q[0] = 0
        if(*remainder & (1 << 31)){
            // update quotient lsb
            *quotient <<= 1;
        }
        // else q[0] = 1
        else{
            *quotient = (*quotient << 1) | 1;
        }

    if (*remainder & (1 << 31)){
        *remainder +=  divisor;
    }
    
    }
}

// TODO: Implement test cases
// Function to run a test case
void runTestCase(uint32_t dividend, uint32_t divisor) {
    // TODO: Call restoringDivision and verify the results
    uint32_t quotient = 0;
    uint32_t remainder = 0;
    nonRestoringDivision(dividend, divisor, &remainder, &quotient);
    // handling division by zero 
    if (divisor == 0){
        printf("Division by Zero Handled correctly \n");
        return;
    }

    uint32_t expected_quotient = dividend / divisor;
    uint32_t expected_remainder = dividend % divisor;

    printf("Dividend: %u, Divisor: %u\n", dividend, divisor);
    printf("Computed Quotient: %u, Computed Remainder: %u\n", quotient, remainder);
    printf("Expected Quotient: %u, Expected Remainder: %u\n", expected_quotient, expected_remainder);

    if (quotient == expected_quotient && remainder == expected_remainder) {
        printf("Test case passed.\n");
    } else {
        printf("Test case failed.\n");
    }
    printf("\n");
}

// Main function to run multiple test cases
int main() {
    // TODO: Implement multiple test cases
    // Seed the random number generator
    srand(time(NULL));
    // runTestCase(dividend, divisor)
    runTestCase(0, 15);
    runTestCase(20, 3);
    runTestCase(1, 20);
    runTestCase(100, 10);
    runTestCase(4294967295, 1);
    runTestCase(1234567890, 987654321);
    runTestCase(11, 3);  // Added the original test case you provided
    runTestCase(( rand() % 100000 ), ( rand() % 1000 ) );
    // division by zero
    runTestCase(20, 0);
    // runTestCase(0, 0);
    return 0;
}


